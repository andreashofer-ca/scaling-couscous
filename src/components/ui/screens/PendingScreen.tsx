import { BusyIndicator } from '@ui5/webcomponents-react'
import { FC } from 'react'
import BusyIndicatorSize from '@ui5/webcomponents/dist/types/BusyIndicatorSize.js'
import { PendingScreenProps } from 'interfaces/properties'
import { BUSY_INDICATOR_DELAY } from 'shared/constants'

const PendingScreen: FC<PendingScreenProps> = ({ title, description }) => (
  <BusyIndicator
    delay={BUSY_INDICATOR_DELAY}
    active
    text={description}
    title={title}
    size={BusyIndicatorSize.L}
    className="full-width-busy-indicator"
  />
)

export default PendingScreen
