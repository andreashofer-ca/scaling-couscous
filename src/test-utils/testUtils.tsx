import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from '@ui5/webcomponents-react'
import React, { ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { ConfigContext } from 'hooks/config/useConfig'
import config from 'test/config'

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })
  return (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider>
        <ConfigContext.Provider value={config}>
          <BrowserRouter>{children}</BrowserRouter>
        </ConfigContext.Provider>
      </ThemeProvider>
    </QueryClientProvider>
  )
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllTheProviders, ...options })

const AllTheProvidersWithoutBrowserRouter = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })
  return (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider>
        <ConfigContext.Provider value={config}>{children}</ConfigContext.Provider>
      </ThemeProvider>
    </QueryClientProvider>
  )
}

const customRenderWithoutBrowserRouter = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllTheProvidersWithoutBrowserRouter, ...options })

export * from '@testing-library/react'
export { customRender as render }

export { customRenderWithoutBrowserRouter as renderWithoutRouter }
