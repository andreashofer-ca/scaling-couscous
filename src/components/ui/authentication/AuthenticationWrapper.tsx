import { useTranslation } from 'react-i18next'
import { useAuth } from 'react-oidc-context'

import { ReactElement, JSXElementConstructor } from 'react'
import PendingScreen from 'components/ui/screens/PendingScreen'

interface Props {
  children: ReactElement<any, string | JSXElementConstructor<any>>
}

const AuthenticationWrapper = ({
  children,
}: Props): ReactElement<any, string | JSXElementConstructor<any>> => {
  const {
    isLoading,
    user,
    activeNavigator,
    signinRedirect,
    events,
    signoutRedirect,
    signinSilent,
  } = useAuth()
  const { t } = useTranslation()

  events.addSilentRenewError(() => {
    signinSilent().catch(() => signoutRedirect())
  })
  events.addAccessTokenExpired(() => {
    signinSilent().catch(() => signoutRedirect())
  })

  const pendingScreen = (
    <PendingScreen
      title={t('app.components.login.title')}
      description={t('app.components.login.signin.description')}
    />
  )

  if (activeNavigator === 'signinSilent' || activeNavigator === 'signoutRedirect') {
    return pendingScreen
  }

  if (user && !user.expired) {
    return children
  }

  if (isLoading) {
    return pendingScreen
  }

  localStorage.redirectUri = window.location.href
  void signinRedirect()
  return children
}

export default AuthenticationWrapper
