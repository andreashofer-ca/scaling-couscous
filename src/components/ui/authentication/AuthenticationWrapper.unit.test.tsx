import { useAuth } from 'react-oidc-context'
import { Mock, vi, describe, beforeEach, afterEach, test, expect, it } from 'vitest'
import PendingScreen from 'components/ui/screens/PendingScreen'
import { render, screen } from 'test-utils/testUtils'
import AuthenticationWrapper from 'components/ui/authentication/AuthenticationWrapper'

vi.mock('components/ui/screens/PendingScreen')
const mockedPendingScreen = PendingScreen as Mock
vi.mock('react-oidc-context')
const mockedUseAuth = useAuth as Mock
const mockSigninRedirect = vi.fn()
const mockSignoutRedirect = vi.fn()
const mockSigninSilent = vi.fn()
const mockAddSilentRenewError = vi.fn()
const mockAddAccessTokenExpired = vi.fn()

const userData = {
  access_token: '123',
  profile: {
    given_name: 'Test',
    family_name: 'Test',
    name: 'Test Test',
  },
}

describe('AuthenticationWrapper', () => {
  beforeEach(() => {
    mockedPendingScreen.mockReturnValue(<div data-testid="PendingScreen" />)
  })

  afterEach(() => {
    vi.clearAllMocks()
  })

  test('when the user is not authenticated', () => {
    mockedUseAuth.mockReturnValue({
      isLoading: false,
      user: null,
      signinRedirect: mockSigninRedirect,
      signoutRedirect: mockSignoutRedirect,
      signinSilent: mockSigninSilent,
      events: {
        addSilentRenewError: mockAddSilentRenewError,
        addAccessTokenExpired: mockAddAccessTokenExpired,
      },
    })

    render(
      <AuthenticationWrapper>
        <div data-testid="Children" />
      </AuthenticationWrapper>,
    )

    expect(mockSigninRedirect).toHaveBeenCalled()
    expect(localStorage.redirectUri).toEqual(window.location.href)
  })
  test('when the user is authenticated', () => {
    mockedUseAuth.mockReturnValue({
      isLoading: false,
      user: userData,
      signinRedirect: mockSigninRedirect,
      signoutRedirect: mockSignoutRedirect,
      signinSilent: mockSigninSilent,
      events: {
        addSilentRenewError: mockAddSilentRenewError,
        addAccessTokenExpired: mockAddAccessTokenExpired,
      },
    })

    render(
      <AuthenticationWrapper>
        <div data-testid="Children" />
      </AuthenticationWrapper>,
    )

    const children = screen.getByTestId('Children')
    expect(children).toBeInTheDocument()
  })
  test('when the authentication is in progress', () => {
    mockedUseAuth.mockReturnValue({
      isLoading: true,
      user: null,
      signinRedirect: mockSigninRedirect,
      signoutRedirect: mockSignoutRedirect,
      signinSilent: mockSigninSilent,
      events: {
        addSilentRenewError: mockAddSilentRenewError,
        addAccessTokenExpired: mockAddAccessTokenExpired,
      },
    })

    const { container, getByTestId } = render(
      <AuthenticationWrapper>
        <div data-testid="Children" />
      </AuthenticationWrapper>,
    )

    expect(container).toMatchSnapshot()
    expect(getByTestId('PendingScreen')).toBeInTheDocument()
  })

  test('when the activeNavigator is in signinSilent', () => {
    mockedUseAuth.mockReturnValue({
      isLoading: true,
      activeNavigator: 'signinSilent',
      user: null,
      signinRedirect: mockSigninRedirect,
      signoutRedirect: mockSignoutRedirect,
      signinSilent: mockSigninSilent,
      events: {
        addSilentRenewError: mockAddSilentRenewError,
        addAccessTokenExpired: mockAddAccessTokenExpired,
      },
    })

    const { getByTestId } = render(
      <AuthenticationWrapper>
        <div data-testid="Children" />
      </AuthenticationWrapper>,
    )

    expect(getByTestId('PendingScreen')).toBeInTheDocument()
  })

  test('when the activeNavigator is in signoutRedirect', () => {
    mockedUseAuth.mockReturnValue({
      isLoading: true,
      activeNavigator: 'signoutRedirect',
      user: null,
      signinRedirect: mockSigninRedirect,
      signoutRedirect: mockSignoutRedirect,
      signinSilent: mockSigninSilent,
      events: {
        addSilentRenewError: mockAddSilentRenewError,
        addAccessTokenExpired: mockAddAccessTokenExpired,
      },
    })

    const { getByTestId } = render(
      <AuthenticationWrapper>
        <div data-testid="Children" />
      </AuthenticationWrapper>,
    )

    expect(getByTestId('PendingScreen')).toBeInTheDocument()
  })
})

describe('when the token expires', () => {
  beforeEach(() => {
    mockedUseAuth.mockReturnValue({
      isLoading: true,
      user: userData,
      signinRedirect: mockSigninRedirect,
      signoutRedirect: mockSignoutRedirect,
      signinSilent: mockSigninSilent,
      events: {
        addSilentRenewError: mockAddSilentRenewError,
        addAccessTokenExpired: mockAddAccessTokenExpired,
      },
    })
  })

  afterEach(() => {
    vi.clearAllMocks()
  })

  describe('for a successful refresh', () => {
    it('should call silentSignin for a refresh', async () => {
      mockSigninSilent.mockResolvedValue({})

      render(
        <AuthenticationWrapper>
          <div data-testid="Children" />
        </AuthenticationWrapper>,
      )

      expect(mockAddAccessTokenExpired).toHaveBeenCalled()
      const callback = mockAddAccessTokenExpired.mock.calls[0][0]
      await callback()

      expect(mockSigninSilent).toHaveBeenCalled()
      expect(mockSignoutRedirect).not.toHaveBeenCalled()
    })
  })
  describe('for an unsuccessful refresh', () => {
    it('should sign out', async () => {
      mockSigninSilent.mockRejectedValue(new Error('test message error'))

      render(
        <AuthenticationWrapper>
          <div data-testid="Children" />
        </AuthenticationWrapper>,
      )

      expect(mockAddAccessTokenExpired).toHaveBeenCalled()
      const callback = mockAddAccessTokenExpired.mock.calls[0][0]
      await callback()

      // Ensure all promises are resolved
      await new Promise(setImmediate)

      expect(mockSigninSilent).toHaveBeenCalled()
      expect(mockSignoutRedirect).toHaveBeenCalled()
    })
  })
})

describe('when is the silentRenew error', () => {
  beforeEach(() => {
    mockedUseAuth.mockReturnValue({
      isLoading: true,
      user: userData,
      signinRedirect: mockSigninRedirect,
      signoutRedirect: mockSignoutRedirect,
      signinSilent: mockSigninSilent,
      events: {
        addSilentRenewError: mockAddSilentRenewError,
        addAccessTokenExpired: mockAddAccessTokenExpired,
      },
    })
  })

  afterEach(() => {
    vi.clearAllMocks()
  })

  describe('for a successful signinSilent', () => {
    it('should call silentSignin once again', async () => {
      mockSigninSilent.mockResolvedValue({})

      render(
        <AuthenticationWrapper>
          <div data-testid="Children" />
        </AuthenticationWrapper>,
      )

      expect(mockAddSilentRenewError).toHaveBeenCalled()
      const callback = mockAddSilentRenewError.mock.calls[0][0]
      await callback()

      expect(mockSigninSilent).toHaveBeenCalled()
      expect(mockSignoutRedirect).not.toHaveBeenCalled()
    })
  })
  describe('for an unsuccessful signinSilent', () => {
    it('should sign out', async () => {
      mockSigninSilent.mockRejectedValue(new Error('test message error'))

      render(
        <AuthenticationWrapper>
          <div data-testid="Children" />
        </AuthenticationWrapper>,
      )

      expect(mockAddSilentRenewError).toHaveBeenCalled()
      const callback = mockAddSilentRenewError.mock.calls[0][0]
      await callback()

      // Ensure all promises are resolved
      await new Promise(setImmediate)

      expect(mockSigninSilent).toHaveBeenCalled()
      expect(mockSignoutRedirect).toHaveBeenCalled()
    })
  })
})
