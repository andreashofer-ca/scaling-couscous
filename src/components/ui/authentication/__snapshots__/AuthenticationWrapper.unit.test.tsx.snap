// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthenticationWrapper > when the authentication is in progress 1`] = `
<div>
  <div
    data-testid="PendingScreen"
  />
</div>
`;
